src/arch/src/aarch64/gic/dist_regs.rs

```diff
@@ -0,0 +1,220 @@
+const GICD_IROUTER: u32 = 0x6000;
+
+// All or at least the registers we are interested in are 32 bit, so
+// we use a constant for size(u32).
+const REG_SIZE: u8 = 4;
+
+/// This is how we represent the registers of the vgic's distributor.
+/// Some of the distributor register )(i.e GICD_STATUSR) are simple
+/// registers (i.e they are associated to a 32 bit value).
+/// However, there are other registers that have variable lengths since
+/// they dedicate some of the 32 bits to some specific interrupt. So, their length
+/// depends on the number of interrupts (i.e the ones that are represented as GICD_REG<n>)
+/// in the documentation mentioned above.
+struct DistReg {
+    /// Offset from distributor address.
+    base: u32,
+    /// Bits per interrupt.
+    /// Relevant for registers that DO share IRQs.
+    bpi: u8,
+    /// Length of the register.
+    /// Relevant for registers that DO NOT share IRQs.
+    length: u16,
+}
+
+impl DistReg {
+    fn compute_reg_range(&self, fd: &DeviceFd) -> Result<(u32, u32)> {
+        let base = self.base + REG_SIZE as u32 * self.bpi as u32;
+        let mut end = base;
+        let num_irq = get_num_interrupts(fd)?;
+
+        if self.length > 0 {
+            // This is the single type register (i.e one that is not DIST_X<n>) and for which
+            // the bpi is 0.
+            // Look in the kernel for REGISTER_DESC_WITH_LENGTH.
+            end = base + self.length as u32;
+        }
+        if self.bpi > 0 {
+            // This is the type of register that takes into account the number of interrupts
+            // that the model has. It is also the type of register where
+            // a register relates to multiple interrupts.
+            end = base + (self.bpi as u32 * (num_irq - super::super::layout::IRQ_BASE) / 8);
+            if self.bpi as u32 * (num_irq - super::super::layout::IRQ_BASE) % 8 > 0 {
+                end += REG_SIZE as u32;
+            }
+        }
+        Ok((base, end))
+    }
+}
+
+// Creates a vgic distributor register.
+macro_rules! VGIC_DIST_REG {
+    ($base:expr, $bpi:expr, $length:expr) => {
+        DistReg {
+            base: $base,
+            bpi: $bpi,
+            length: $length,
+        }
+    };
+}
+
+// List with relevant distributor registers that we will be restoring.
+// Order is taken from qemu.
+static VGIC_DIST_REGS: &[DistReg] = &[
+    VGIC_DIST_REG!(GICD_STATUSR, 0, 4),
+    VGIC_DIST_REG!(GICD_ICENABLER, 1, 0),
+    VGIC_DIST_REG!(GICD_ISENABLER, 1, 0),
+    VGIC_DIST_REG!(GICD_IGROUPR, 1, 0),
+    VGIC_DIST_REG!(GICD_IROUTER, 64, 0),
+    VGIC_DIST_REG!(GICD_ICFGR, 2, 0),
+    VGIC_DIST_REG!(GICD_ICPENDR, 1, 0),
+    VGIC_DIST_REG!(GICD_ISPENDR, 1, 0),
+    VGIC_DIST_REG!(GICD_ICACTIVER, 1, 0),
+    VGIC_DIST_REG!(GICD_ISACTIVER, 1, 0),
+    VGIC_DIST_REG!(GICD_IPRIORITYR, 8, 0),
+];
+
+#[allow(clippy::trivially_copy_pass_by_ref)]
+fn kvm_set_dist_regs(fd: &DeviceFd, offset: u32, val: &u32) -> Result<()> {
```

Comment:
Can we move this into the impl for `DistReg` ? Seems dist reg logic.

Also since it sets only one reg, should we remove the plural? `kvm_set_dist_reg` ?

The same comments for `kvm_get_dist_regs`.